parser/prism/Translator.cc

```diff
@@ -69,33 +71,39 @@ std::unique_ptr<SorbetAssignmentNode> Translator::translateOpAssignment(pm_node_
     } else if constexpr (std::is_same_v<PrismAssignmentNode, pm_constant_operator_write_node> ||
                          std::is_same_v<PrismAssignmentNode, pm_constant_and_write_node> ||
                          std::is_same_v<PrismAssignmentNode, pm_constant_or_write_node>) {
-        auto *nameLoc = &node->name_loc;
-        auto name = parser.resolveConstant(node->name);
-        lhs = make_unique<SorbetLHSNode>(parser.translateLocation(nameLoc), nullptr, gs.enterNameConstant(name));
+        // Handle operator assignment to a "plain" constant, like `A += 1`
+        lhs = translateConst<PrismAssignmentNode, parser::ConstLhs>(node);
     } else if constexpr (std::is_same_v<PrismAssignmentNode, pm_constant_path_operator_write_node> ||
                          std::is_same_v<PrismAssignmentNode, pm_constant_path_and_write_node> ||
                          std::is_same_v<PrismAssignmentNode, pm_constant_path_or_write_node>) {
-        auto isAssignment = true;
-        lhs = translateConstantPath(node->target, isAssignment);
+        // Handle operator assignment to a constant path, like `A::B::C += 1` or `::C += 1`
+        lhs = translateConst<pm_constant_path_node, parser::ConstLhs>(node->target);
     } else if constexpr (std::is_same_v<SorbetLHSNode, parser::Send>) {
+        // Handle operator assignment to the result of a method call, like `a.b += 1`
         auto name = parser.resolveConstant(node->read_name);
         auto receiver = translate(node->receiver);
         auto *message_loc = &node->message_loc;
         lhs = make_unique<parser::Send>(parser.translateLocation(loc), std::move(receiver), gs.enterNameUTF8(name),
                                         parser.translateLocation(message_loc), NodeVec{});
     } else {
+        // Handle regular assignment to any other kind of LHS.
         auto *nameLoc = &node->name_loc;
         auto name = parser.resolveConstant(node->name);
         lhs = make_unique<SorbetLHSNode>(parser.translateLocation(nameLoc), gs.enterNameUTF8(name));
     }
 
     if constexpr (std::is_same_v<SorbetAssignmentNode, parser::OpAsgn>) {
+        // `OpAsgn` assign needs more information about the specific operator here, so it gets special handling here.
         auto *opLoc = &node->binary_operator_loc;
         auto op = parser.resolveConstant(node->binary_operator);
 
         return make_unique<parser::OpAsgn>(parser.translateLocation(loc), std::move(lhs), gs.enterNameUTF8(op),
                                            parser.translateLocation(opLoc), std::move(rhs));
     } else {
+        // `AndAsgn` and `OrAsgn` are specific to a single operator, so don't need any extra information like `OpAsgn`.
+        static_assert(std::is_same_v<SorbetAssignmentNode, parser::AndAsgn> ||
+                      std::is_same_v<SorbetAssignmentNode, parser::OrAsgn>);
```

Comment:
This isn't strictly necessary, since all possible types of `SorbetAssignmentNode` are enumerated above:

https://github.com/Shopify/sorbet/blob/724cde445e0ffd76470d514d13bc4e872ed8b2a5/parser/prism/Translator.cc#L49-L52

But it's still nice to see locally what the possible types are that can end up here.